File: couchbase-entity/src/main/java/com/kaufland/generation/WrapperGeneration.java
package com.kaufland.generation;

import com.kaufland.model.entity.BaseEntityHolder;
import com.kaufland.model.entity.WrapperEntityHolder;
import com.kaufland.model.field.CblBaseFieldHolder;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.lang.model.element.Modifier;

public class WrapperGeneration {

    public JavaFile generateModel(WrapperEntityHolder holder) {

        TypeSpec.Builder typeBuilder = TypeSpec.classBuilder(holder.getEntitySimpleName()).
                addModifiers(Modifier.PUBLIC).
                addField(CblDefaultGeneration.field()).
                addMethods(create(holder)).
                addField(TypeUtil.createMapStringObject(), "mDoc", Modifier.PRIVATE).
                addMethod(contructor()).
                superclass(TypeName.get(holder.getSourceElement().asType()));

        for (CblBaseFieldHolder fieldHolder : holder.getAllFields()) {

            typeBuilder.addFields(fieldHolder.createFieldConstant());
            typeBuilder.addMethod(fieldHolder.getter(null, false));

            MethodSpec setter = fieldHolder.setter(null, holder.getEntityTypeName(), false);
            if (setter != null) {
                typeBuilder.addMethod(setter);
            }
        }

        typeBuilder.addStaticBlock(CblDefaultGeneration.staticInitialiser(holder));
        typeBuilder.addMethod(new RebindMethodGeneration().generate(false));
        typeBuilder.addMethods(fromMap(holder));
        typeBuilder.addMethods(toMap(holder));
        typeBuilder.addMethods(new TypeConversionMethodsGeneration().generate());

        return JavaFile.builder(holder.getPackage(), typeBuilder.build()).
                build();

    }

    private List<MethodSpec> toMap(BaseEntityHolder holder) {
        CodeBlock nullCheck = CodeBlock.builder().
                beginControlFlow("if(obj == null)").
                addStatement("return null").
                endControlFlow().
                build();

        return Arrays.asList(MethodSpec.methodBuilder("toMap").addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                                     addParameter(holder.getEntityTypeName(), "obj").
                                     returns(TypeUtil.createMapStringObject()).
                                     addCode(nullCheck).
                                     addStatement("$T result = new $T()", TypeUtil.createHashMapStringObject(), TypeUtil.createHashMapStringObject())
                                     .addStatement("result.putAll(mDocDefaults)").
                                             addStatement("result.putAll(obj.mDoc)").
                                             addStatement("return result").
                                             build(),

                             MethodSpec.methodBuilder("toMap").addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                                     addParameter(ParameterizedTypeName.get(ClassName.get(List.class), holder.getEntityTypeName()), "obj").
                                     returns(TypeUtil.createListWithMapStringObject()).
                                     addCode(nullCheck).
                                     addStatement("$T result = new $T()", TypeUtil.createListWithMapStringObject(), TypeUtil.createArrayListWithMapStringObject()).
                                     addCode(CodeBlock.builder().beginControlFlow("for($N entry : obj)", holder.getEntitySimpleName()).
                                             addStatement("$T temp = new $T()", TypeUtil.createHashMapStringObject(), TypeUtil.createHashMapStringObject()).
                                             addStatement("temp.putAll((($N)entry).toMap(entry))", holder.getEntitySimpleName()).
                                             addStatement("temp.putAll(mDocDefaults)").
                                             addStatement("result.add(temp)", holder.getEntitySimpleName()).
                                             endControlFlow().
                                             build()).
                                     addStatement("return result").
                                     build());
    }

    private List<MethodSpec> fromMap(BaseEntityHolder holder) {
        CodeBlock nullCheck = CodeBlock.builder().
                beginControlFlow("if(obj == null)").
                addStatement("return null").
                endControlFlow().
                build();

        return Arrays.asList(MethodSpec.methodBuilder("fromMap").addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                                     addParameter(TypeUtil.createMapStringObject(), "obj").
                                     returns(holder.getEntityTypeName()).
                                     addCode(nullCheck).
                                     addStatement("return new $T(obj)", holder.getEntityTypeName()).
                                     build(),

                             MethodSpec.methodBuilder("fromMap").addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                                     addParameter(ParameterizedTypeName.get(ClassName.get(List.class), TypeUtil.createMapStringObject()), "obj").
                                     returns(ParameterizedTypeName.get(ClassName.get(List.class), holder.getEntityTypeName())).
                                     addCode(nullCheck).
                                     addStatement("$T result = new $T()", ParameterizedTypeName.get(ClassName.get(List.class), holder.getEntityTypeName()), ParameterizedTypeName.get(ClassName.get(ArrayList.class), holder.getEntityTypeName())).
                                     addCode(CodeBlock.builder().beginControlFlow("for($T entry : obj)", TypeUtil.createMapStringObject()).
                                             addStatement("result.add(new $N(entry))", holder.getEntitySimpleName()).
                                             endControlFlow().
                                             build()).
                                     addStatement("return result").
                                     build()
        );
    }

    private MethodSpec contructor() {
        return MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC).
                addParameter(TypeUtil.createMapStringObject(), "doc").
                addStatement("rebind(doc)").
                build();
    }

    private List<MethodSpec> create(WrapperEntityHolder holder) {

        return Arrays.asList(
                MethodSpec.methodBuilder("create").
                        addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                        addParameter(TypeUtil.createMapStringObject(), "doc").
                        addStatement("return new $N (doc)",
                                     holder.getEntitySimpleName()).
                        returns(holder.getEntityTypeName()).
                        build(),
                MethodSpec.methodBuilder("create").
                        addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                        addStatement("return new $N (new $T())",
                                     holder.getEntitySimpleName(), TypeUtil.createHashMapStringObject()).
                        returns(holder.getEntityTypeName()).
                        build()
        );
    }

}

File: couchbase-entity/src/main/java/com/kaufland/generation/WrapperGeneration.kt
package com.kaufland.generation

import com.kaufland.model.entity.BaseEntityHolder
import com.kaufland.model.entity.WrapperEntityHolder
import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.*
import java.util.*

class WrapperGeneration {

    fun generateModel(holder: WrapperEntityHolder, useSuspend: Boolean): FileSpec {

        val companionSpec = TypeSpec.companionObjectBuilder()

        val builderBuilder = BuilderClassGeneration.generateBaseBuilder(holder)

        val typeBuilder = TypeSpec.classBuilder(holder.entitySimpleName)
                .addSuperinterface(TypeUtil.mapSupport())
                .addModifiers(KModifier.PUBLIC)
                .addSuperinterface(holder.interfaceTypeName)
                .addFunction(EnsureTypesGeneration.ensureTypes(holder, true))
                .addFunction(CblDefaultGeneration.addDefaults(holder, true))
                .addFunction(CblConstantGeneration.addConstants(holder, true))
                .addFunction(MapSupportGeneration.toMap(holder))
                .addProperty(PropertySpec.builder("mDoc", TypeUtil.mutableMapStringAnyNullable()).addModifiers(KModifier.PRIVATE).mutable().initializer("%T()", TypeUtil.linkedHashMapStringAnyNullable()).build())
                .addFunction(constructorMap())
                .addFunction(constructorDefault())
                .superclass(holder.sourceElement!!.asType().asTypeName())
                .addFunction(BuilderClassGeneration.generateBuilderFun())

        holder.deprecated?.addDeprecated(typeBuilder)

        holder.docId?.let {
            companionSpec.addFunction(it.companionFunction(holder))
            typeBuilder.addFunction(it.buildExpectedDocId(holder))
            typeBuilder.addSuperinterface(TypeUtil.iDocId())
        }
        if (holder.comment.isNotEmpty()) {
            typeBuilder.addKdoc(holder.comment.joinToString(separator = "\n"))
        }

        for (baseModelHolder in holder.basedOn) {
            typeBuilder.addSuperinterface(baseModelHolder.interfaceTypeName)
        }

        if(holder.modifierOpen){
            typeBuilder.addModifiers(KModifier.OPEN)
        }

        for (fieldHolder in holder.allFields) {

            companionSpec.addProperties(fieldHolder.createFieldConstant())
            typeBuilder.addProperty(fieldHolder.property(null, holder.abstractParts, false, holder.deprecated))
            fieldHolder.builderSetter(null, holder.`package`, holder.entitySimpleName, false)?.let {
                builderBuilder.addFunction(it)
            }
        }

        companionSpec.addFunctions(fromMap(holder))
        companionSpec.addFunctions(toMap(holder))
        companionSpec.addFunctions(create(holder))
        typeBuilder.addType(companionSpec.build())
        typeBuilder.addFunction(RebindMethodGeneration().generate(false))
        typeBuilder.addFunctions(TypeConversionMethodsGeneration(useSuspend).generate())
        typeBuilder.addType(builderBuilder.build())

        return FileSpec.get(holder.`package`, typeBuilder.build())

    }

    private fun toMap(holder: BaseEntityHolder): List<FunSpec> {
        val nullCheck = CodeBlock.builder().beginControlFlow("if(obj == null)").addStatement("return mutableMapOf()").endControlFlow().build()
        val nullCheckList = CodeBlock.builder().beginControlFlow("if(obj == null)").addStatement("return listOf()").endControlFlow().build()

        return Arrays.asList(FunSpec.builder("toMap").addModifiers(KModifier.PUBLIC)
                .addParameter("obj", holder.entityTypeName.copy(nullable = true)).returns(TypeUtil.mutableMapStringAny())
                .addAnnotation(JvmStatic::class)
                .addCode(nullCheck).addStatement("var result = mutableMapOf<%T,%T>()", TypeUtil.string(), TypeUtil.any())
                .beginControlFlow("obj.mDoc.forEach")
                .beginControlFlow("if(it.value != null)").addStatement("result[it.key] = it.value!!").endControlFlow()
                .endControlFlow()
                .addStatement("return result").build(),

                FunSpec.builder("toMap").addModifiers(KModifier.PUBLIC)
                        .addParameter("obj", TypeUtil.list(holder.entityTypeName).copy(nullable = true)).addAnnotation(JvmStatic::class)
                        .returns(TypeUtil.listWithMutableMapStringAny()).addCode(nullCheckList)
                        .addStatement("var result = %T()", TypeUtil.arrayListWithMutableMapStringAny())
                        .addCode(CodeBlock.builder()
                                .beginControlFlow("for(entry in obj)")
                                .addStatement("var temp = mutableMapOf<%T,%T>()", TypeUtil.string(), TypeUtil.any())
                                .addStatement("temp.putAll(%N.toMap(entry)!!)", holder.entitySimpleName)
                                .addStatement("result.add(temp)", holder.entitySimpleName).endControlFlow().build())
                        .addStatement("return result").build())
    }

    private fun fromMap(holder: BaseEntityHolder): List<FunSpec> {
        val nullCheck = CodeBlock.builder().beginControlFlow("if(obj == null)").addStatement("return null").endControlFlow().build()

        return Arrays.asList(FunSpec.builder("fromMap").addModifiers(KModifier.PUBLIC)
                .addParameter("obj", TypeUtil.mutableMapStringAnyNullable().copy(nullable = true)).addAnnotation(JvmStatic::class)
                .returns(holder.entityTypeName.copy(nullable = true)).addCode(nullCheck)
                .addStatement("return %T(obj)", holder.entityTypeName).build(),

                FunSpec.builder("fromMap").addModifiers(KModifier.PUBLIC).addAnnotation(JvmStatic::class)
                        .addParameter("obj", TypeUtil.listWithMutableMapStringAnyNullable().copy(nullable = true))
                        .returns(TypeUtil.list(holder.entityTypeName).copy(nullable = true)).addCode(nullCheck)
                        .addStatement("var result = %T()", TypeUtil.arrayList(holder.entityTypeName))
                        .addCode(CodeBlock.builder().beginControlFlow("for(entry in obj)")
                                .addStatement("result.add(%N(entry))", holder.entitySimpleName)
                                .endControlFlow().build()).addStatement("return result").build()
        )
    }

    private fun constructorMap(): FunSpec {
        return FunSpec.constructorBuilder().addModifiers(KModifier.PUBLIC).addParameter("doc", TypeUtil.mutableMapStringAnyNullable()).addStatement("rebind(ensureTypes(doc))").build()
    }

    private fun constructorDefault(): FunSpec {
        // Add default constructor to allow property-based creators in Jackson
        return FunSpec.constructorBuilder().addModifiers(KModifier.PUBLIC).callThisConstructor("mutableMapOf()") .build()
    }

    private fun create(holder: WrapperEntityHolder): List<FunSpec> {

        return Arrays.asList(
                FunSpec.builder("create").addModifiers(KModifier.PUBLIC).addParameter("doc", TypeUtil.mutableMapStringAnyNullable()).addAnnotation(JvmStatic::class).addStatement("return %N(doc)",
                        holder.entitySimpleName).returns(holder.entityTypeName).build(),
                FunSpec.builder("create").addModifiers(KModifier.PUBLIC).addAnnotation(JvmStatic::class).addStatement("return %N(%T())",
                        holder.entitySimpleName, TypeUtil.hashMapStringAnyNullable()).returns(holder.entityTypeName).build()
        )
    }

}
